### PR TITLE
Initialize ImagePipelineFactory in ImageLoadedModule

### LINKED ISSUE #8397: ImagePipelineFactory was not initialized!
1. Checkout the source code
2. Build and Install UIExplorer to Android Devices
3. The first time to launch UIExplorer app

image

But after `RELOAD`, everything works fine.

### PR DESCRIPTION
**Pull request by leeight**
FIX #8397 #8114 

In the Examples/UIExplorer/ImageExample.js, Image.prefetch was called,
but ImagePipelineFactory was not initialized.
We should make ImagePipelineFactory available before use it.

### CODE REVIEW

**Comment by facebook-github-bot-5** [comment]
By analyzing the blame information on this pull request, we identified @ide and @cornedor to be potential reviewers.

**Comment by janicduplessis** [comment]
I guess there is some sort of race condition with fresco initialization, it is normally done [here](https://github.com/leeight/react-native/blob/206d2d5d714cdb7abfbd0b717e08dc9f5395790b/ReactAndroid/src/main/java/com/facebook/react/modules/fresco/FrescoModule.java#L65). Maybe we should move the initialization code so some static class like `FrescoHelpers` and use it to make sure it is initialized in both modules.

**Comment by facebook-github-bot-9** [comment]
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request we identified @ide as a potential reviewer. Could you take a look please or cc someone with more context?

**Comment by leeight** [comment]
@janicduplessis How about this?

```java
public class FrescoHelpers {
    static void initialize(context) {
        Fresco.initialize(context);
    }
}
```

And the call `FrescoHelpers.initialize()` in ImageLoaderModule.java and FrescoModule.java?

**Comment by chamchamgo** [comment]
I have a work-around if you only want to use UIExplorer  (modify ImageExample.js under UIExplorer  root)  :
`...
//var prefetchTask = Image.prefetch(IMAGE_PREFETCH_URL);
var prefetchTask;  //chamchamgo

var NetworkImageCallbackExample = React.createClass({
  getInitialState: function() {
		prefetchTask = Image.prefetch(IMAGE_PREFETCH_URL);  //chamchamgo
...`

**Comment by facebook-github-bot-8** [comment]
@leeight updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8398/files/206d2d5d714cdb7abfbd0b717e08dc9f5395790b..386d071bc346bbffb62a283f746a80e6ed4e294c)

**Comment by facebook-github-bot-9** [comment]
@leeight updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8398/files/386d071bc346bbffb62a283f746a80e6ed4e294c..730e5f74382631d9681293ee88004452c4a8c11c)

**Comment by leeight** [comment]
Ping @janicduplessis @mkonicek

**Comment by facebook-github-bot-6** [comment]
@leeight updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8398/files/730e5f74382631d9681293ee88004452c4a8c11c..f68423a55b3cb757adf4feba4c28e139299b9caa)

**Comment by facebook-github-bot-6** [comment]
@leeight updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8398/files/f68423a55b3cb757adf4feba4c28e139299b9caa..534c6c5bd80abf9965e99c7f0f92b1f577d91e53)

**Comment by facebook-github-bot-3** [comment]
@leeight updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8398/files/534c6c5bd80abf9965e99c7f0f92b1f577d91e53..3c3798a6d66ef2d3155fa5dd376af4c17a2e4eb5)

**Comment by facebook-github-bot-9** [comment]
@leeight updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8398/files/3c3798a6d66ef2d3155fa5dd376af4c17a2e4eb5..61e1015ff4771ba00566d5aba5021a613a5f86d7)

**Comment by leeight** [comment]
Ping @janicduplessis @ide

**Comment by facebook-github-bot-5** [comment]
@leeight updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8398/files/61e1015ff4771ba00566d5aba5021a613a5f86d7..ea9ccbe243be280457dbceae303da62c33f9a549)

**Comment by facebook-github-bot-2** [comment]
@leeight updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8398/files/ea9ccbe243be280457dbceae303da62c33f9a549..22c8a9aef9e3328c21fb2d9aa76d0569b193b3e7)

File: ReactAndroid/src/main/java/com/facebook/react/modules/image/ImageLoaderModule.java [hunk 1]
```diff
@@ -78,6 +85,9 @@ public void getSize(
       return;
     }
 
+    // Ensure FrescoHelpers.initialize was called.
+    this.initialize();
```
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/modules/image/ImageLoaderModule.java]: Is this safe to call `super.initialize()` extra times?

Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/image/ImageLoaderModule.java]: Oh, github has replies to comments now!

**Comment by janicduplessis** [comment]
@mkonicek Is this fix still needed? There was some talk about the Fresco team fixing it but AFAIK I never saw it land. Related: #9606.

File: ReactAndroid/src/main/java/com/facebook/react/modules/image/ImageLoaderModule.java (hunk 1 above)
Comment (leeight) [inline ReactAndroid/src/main/java/com/facebook/react/modules/image/ImageLoaderModule.java]: @ide `super.initialize` do nothing, so it's safe now.

**Comment by leeight** [comment]
@janicduplessis IMHO, The root cause is `ImageLoader.prefetch` was called before `ImageLoaderModule.initialize` and `FrescoModule.initialize`, so i think it's different with 9606 ?

File: ReactAndroid/src/main/java/com/facebook/react/modules/image/ImageLoaderModule.java (hunk 1 above)
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/modules/image/ImageLoaderModule.java]: Is it always safe according to the API contract? Just because it works doesn't mean it's right.

**Comment by oprisnik** [comment]
Yes this is different. Only initializing Fresco once was implemented in c0de1a7011846e780d9ddfb9d4427b8b17830c0a

File: ReactAndroid/src/main/java/com/facebook/react/modules/image/ImageLoaderModule.java (hunk 1 above)
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/modules/image/ImageLoaderModule.java]: Is this even needed? The module initialize method should always get called before getSize can be called.

**Comment by facebook-github-bot-3** [comment]
@leeight updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8398/files/22c8a9aef9e3328c21fb2d9aa76d0569b193b3e7..5f2e9c1c5a510865d05cb71a618bba13dfa9bf96)

**Comment by facebook-github-bot-9** [comment]
@leeight updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8398/files/5f2e9c1c5a510865d05cb71a618bba13dfa9bf96..83073fa68892ed2a146933d05667c2deb0126a43)

**Comment by leeight** [comment]
@oprisnik `FrescoModule` and `ImageLoaderModule` initialization order is not guaranteed, so i move the initialization logic to `FrescoHelpers`, and call `FrescoHelpers.initialize` in those two modules.

**Comment by oprisnik** [comment]
Just moving this to a helper method seems fine to me. Somebody from React Native should take a closer look and import :)

**Comment by janicduplessis** [comment]
@leeight Changes look good to me, could you just address the inline comments, mostly about making sure we're always passing the same config and the extra initialize calls.

File: ReactAndroid/src/main/java/com/facebook/react/modules/image/ImageLoaderModule.java (hunk 1 above)
Comment (leeight) [inline ReactAndroid/src/main/java/com/facebook/react/modules/image/ImageLoaderModule.java]: I add some breakpoints in Android Studio, launch UIExplorer app, and found `ImageLoader.prefetch` was called before `ImageLoader.initialize`. I haven't found out the root cause, but i think it's worth send another PR.

Comment (leeight) [inline ReactAndroid/src/main/java/com/facebook/react/modules/image/ImageLoaderModule.java]: I add some breakpoints in Android Studio, launch UIExplorer app, and found `ImageLoader.prefetch` was called before `ImageLoader.initialize`. I haven't found out the root cause, but i think it's worth send another PR.

**Comment by facebook-github-bot-5** [comment]
@leeight updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8398/files/83073fa68892ed2a146933d05667c2deb0126a43..65d9a184f99b2096f9700914e0f4bb2c622fa970)

**Comment by facebook-github-bot-4** [comment]
@leeight updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8398/files/65d9a184f99b2096f9700914e0f4bb2c622fa970..3820aa17b2bd33e4861b118ea20e09de42b45a58)

File: ReactAndroid/src/main/java/com/facebook/react/modules/image/ImageLoaderModule.java (hunk 1 above)
Comment (leeight) [inline ReactAndroid/src/main/java/com/facebook/react/modules/image/ImageLoaderModule.java]: @ide Remove the duplicate call `super.initialize()`

**Comment by facebook-github-bot-2** [comment]
@leeight updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8398/files/3820aa17b2bd33e4861b118ea20e09de42b45a58..11642550c39f2bedc540ae0eea4bc4812293c987)

**Comment by facebook-github-bot-7** [comment]
@leeight updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8398/files/11642550c39f2bedc540ae0eea4bc4812293c987..8544129e5e1622203a7ef62cc23bdd221997f09e)

**Comment by leeight** [comment]
@janicduplessis Is there anything else that needs to be updated?

**Comment by janicduplessis** [comment]
@leeight Sorry didn't notice the new github review thing didn't actually post my comment.

**Comment by oprisnik** [comment]
cc @janicduplessis @foghina

File: Examples/UIExplorer/js/ToastAndroidExample.android.js [hunk 2]
```diff
@@ -27,6 +27,7 @@ var React = require('react');
 var ReactNative = require('react-native');
 var {
   StyleSheet,
+  View,
```
Comment (foghina) [inline Examples/UIExplorer/js/ToastAndroidExample.android.js]: All changes in this file seem unrelated?

File: ReactAndroid/src/main/java/com/facebook/react/modules/fresco/FrescoHelpers.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        config = getDefaultConfig(reactContext);
+      }
+
+      // Make sure the SoLoaderShim is configured to use our loader for native libraries.
+      // This code can be removed if using Fresco from Maven rather than from source
+      SoLoaderShim.setHandler(new FrescoHandler());
+      Fresco.initialize(reactContext.getApplicationContext(), config);
+      sHasBeenInitialized = true;
+    }
+    else if (config != null) {
```
Comment (foghina) [inline ReactAndroid/src/main/java/com/facebook/react/modules/fresco/FrescoHelpers.java]: Let's remove the `config != null` check. Trying to initialize multiple times should result in a warning regardless of what config you're trying to use.

File: ReactAndroid/src/main/java/com/facebook/react/modules/fresco/FrescoHelpers.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  public static void initialize(ReactApplicationContext reactContext, @Nullable ImagePipelineConfig config) {
+    if (!hasBeenInitialized()) {
+      if (config == null) {
+        config = getDefaultConfig(reactContext);
+      }
+
+      // Make sure the SoLoaderShim is configured to use our loader for native libraries.
+      // This code can be removed if using Fresco from Maven rather than from source
+      SoLoaderShim.setHandler(new FrescoHandler());
+      Fresco.initialize(reactContext.getApplicationContext(), config);
```
Comment (foghina) [inline ReactAndroid/src/main/java/com/facebook/react/modules/fresco/FrescoHelpers.java]: @oprisnik does fresco make any assumptions as to which thread this is called on? Should we e.g. assert this is always called on the UI thread?

Comment (oprisnik) [inline ReactAndroid/src/main/java/com/facebook/react/modules/fresco/FrescoHelpers.java]: I think you can do this from any thread, so I don't think a check is needed.
There could be an issue if you send a request before initialization finished (e.g. from another thread), but otherwise it should work.

**Comment by leeight** [comment]
I'll update it according this comments, thanks you

**Comment by leeight** [comment]
Updated

**Comment by lacker** [comment]
The circleci errors look real:

    /home/ubuntu/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/image/ImageLoaderModule.java:6    8: error: cannot access com.facebook.react.modules.common.ModuleDataCleaner
        if (!FrescoModule.hasBeenInitialized()) {
                         ^
      class file for com.facebook.react.modules.common.ModuleDataCleaner not found

File: ReactAndroid/src/main/java/com/facebook/react/modules/fresco/FrescoModule.java [hunk 5]
```diff
@@ -63,29 +62,39 @@ public FrescoModule(ReactApplicationContext reactContext) {
...
 
-  @Override
-  public void initialize() {
-    super.initialize();
+  public void initializeFresco() {
     if (!hasBeenInitialized()) {
       // Make sure the SoLoaderShim is configured to use our loader for native libraries.
       // This code can be removed if using Fresco from Maven rather than from source
       SoLoaderShim.setHandler(new FrescoHandler());
-      if (mConfig == null) {
```
Comment (AaaChiuuu) [inline ReactAndroid/src/main/java/com/facebook/react/modules/fresco/FrescoModule.java]: Why was this if case deleted from here and moved up? If setConfig(null) is called multiple we create a new default config multiple times.

File: ReactAndroid/src/main/java/com/facebook/react/modules/fresco/FrescoModule.java [hunk 6]
```diff
@@ -63,29 +62,39 @@ public FrescoModule(ReactApplicationContext reactContext) {
...
-        mConfig = getDefaultConfig(getReactApplicationContext());
-      }
-      Context context = getReactApplicationContext().getApplicationContext();
-      Fresco.initialize(context, mConfig);
+      Fresco.initialize(this.getReactApplicationContext(), mConfig);
```
Comment (AaaChiuuu) [inline ReactAndroid/src/main/java/com/facebook/react/modules/fresco/FrescoModule.java]: Why change the call to initialize with the ReactApplicationContext instead of the ApplicationContext?

File: ReactAndroid/src/main/java/com/facebook/react/modules/fresco/FrescoModule.java [hunk 7]
```diff
@@ -63,29 +62,39 @@ public FrescoModule(ReactApplicationContext reactContext) {
...
       sHasBeenInitialized = true;
-    } else if (mConfig != null) {
-      FLog.w(
-          ReactConstants.TAG,
-          "Fresco has already been initialized with a different config. "
-          + "The new Fresco configuration will be ignored!");
+    } else {
+      FLog.w(ReactConstants.TAG, "Fresco has already been initialized!");
     }
-    mConfig = null;
```
Comment (AaaChiuuu) [inline ReactAndroid/src/main/java/com/facebook/react/modules/fresco/FrescoModule.java]: Why remove null-ing out mConfig? Now we risk a memory leak.

File: ReactAndroid/src/main/java/com/facebook/react/modules/fresco/FrescoModule.java [hunk 8]
```diff
@@ -63,29 +62,39 @@ public FrescoModule(ReactApplicationContext reactContext) {
...
       // This code can be removed if using Fresco from Maven rather than from source
       SoLoaderShim.setHandler(new FrescoHandler());
-      if (mConfig == null) {
-        mConfig = getDefaultConfig(getReactApplicationContext());
-      }
-      Context context = getReactApplicationContext().getApplicationContext();
-      Fresco.initialize(context, mConfig);
+      Fresco.initialize(this.getReactApplicationContext(), mConfig);
       sHasBeenInitialized = true;
-    } else if (mConfig != null) {
```
Comment (AaaChiuuu) [inline ReactAndroid/src/main/java/com/facebook/react/modules/fresco/FrescoModule.java]: I think this case can stay deleted with mConfig being set to null below. Looks like a refactor may have gone wrong.

**Comment by leeight** [comment]
Close it and open a new One.
